src/workerd/io/actor-sqlite.c++

```diff
@@ -169,11 +223,97 @@ void ActorSqlite::onWrite() {
       // rather than after the callback.
       { auto drop = kj::mv(txn); }
 
-      return commitCallback();
+      return commitImpl(kj::mv(precommitAlarmState));
     })));
   }
 }
 
+kj::Promise<void> ActorSqlite::requestScheduledAlarm(kj::Maybe<kj::Date> requestedTime) {
+  // Not using coroutines here, because it's important for correctness in workerd that a
+  // synchronously thrown exception in scheduleRun() can escape synchronously to the caller.
+
+  bool movingAlarmLater = willFireEarlier(lastConfirmedScheduledAlarm, requestedTime);
+  if (movingAlarmLater) {
+    // Since we are setting the alarm to be later, we can update lastConfirmedScheduledAlarm
+    // immediately and still preserve the invariant that the scheduled alarm time is equal to or
+    // earlier than the persisted db alarm value.  Doing the immediate update ensures that
+    // subsequent invocations of commitImpl() will compare against the correct value in their
+    // precommit alarm checks, even if other later-setting requests are still in-flight, without
+    // needing to wait for them to complete.
+    lastConfirmedScheduledAlarm = requestedTime;
```

Comment:
Perhaps `lastConfirmedScheduledAlarm` should actually be called something like... `alarmScheduledNoLaterThan`? Makes clearer why it's safe to update before the call when scheduling later, but only after the call when scheduling earlier.

Reply:
Thanks! It was bugging me that it no longer represented "confirmed" alarms, but I wasn't sure what a good name was.

Reply:
Fixed in d1983124b .